core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java

```diff
@@ -218,8 +214,30 @@ public RelOptPredicateList getPredicates(Project project,
     for (RexNode r : inputInfo.pulledUpPredicates) {
       RexNode r2 = projectPredicate(rexBuilder, input, r, columnsMapped);
       if (!r2.isAlwaysTrue()) {
-        r2 = r2.accept(new RexPermuteInputsShuttle(m, input));
-        projectPullUpPredicates.add(r2);
+        ImmutableBitSet fields = RelOptUtil.InputFinder.bits(r2);
+        // If r2 cannot find input (such as SubQuery),
+        // it will directly return without adjusting mapping.
+        if (fields.isEmpty()) {
+          projectPullUpPredicates.add(r2);
+          continue;
+        }
+        // Using Map<Integer, BitSet>, the result will be
```

Comment:
There is no need to describe alternative designs.
I would change this comment and add it to the declaration of the map itself, above.
I would document the map: keys are field indexes (RexInputRef) that appear in the project, values are sets of field indexes (RexInputRef) that appear in r2.

Reply:
I updated the doc and removed the annotations. I tested locally that both java8 and java17 were successfully and to see if ci can succeed.